core/src/main/java/com/google/zxing/qrcode/encoder/MinimalEncoder.java

```diff
@@ -0,0 +1,778 @@
+    if (DEBUG) {
+        System.err.println("DEBUG Initial situation");
+        printEdges(vertices);
+    }
+    for (int i = 1; i <= inputLength; i++) {
+      for (int j = 0; j < encoders.length; j++) {
+        for (int k = 0; k < 4; k++) {
+          ResultNode minimalEdge = null;
+          if (vertices[i][j][k] != null) {
+            Vector<ResultNode> edges = vertices[i][j][k];
+            if (edges.size() == 1) { //Optimization: if there is only one edge then that one is the minimal one
+              minimalEdge = edges.get(0);
+            } else {
+              int minimalIndex = -1;
+              int minimalSize = Integer.MAX_VALUE;
+              for (int l = 0; l < edges.size(); l++) {
+                ResultNode rn = edges.get(l);
+                if (rn.getSize(true) < minimalSize) {
+                  minimalIndex = l;
+                  minimalSize = rn.getSize(true);
+                }
+              }
+              assert minimalIndex != -1;
+              minimalEdge = edges.get(minimalIndex);
+              edges.clear();
+              edges.add(minimalEdge);
+            }
+            if (i < inputLength) {
+              assert minimalEdge != null;
+              if (DEBUG && minimalEdge != null) {
+                System.err.println("DEBUG processing " + vertexToString(i,minimalEdge) + ". The minimal edge leading to this vertex is " + edgeToString(minimalEdge) + " with a size of " + minimalEdge.getSize(true));
+              }
+              addEdges(version,vertices,i,minimalEdge);
+            }
+          }
+        }
+      }
+      if (DEBUG) {
+        System.err.println("DEBUG situation after adding edges to vertices at position " + i);
+        printEdges(vertices);
+      }
+    }
+    int minimalJ = -1;
+    int minimalK = -1;
+    int minimalSize = Integer.MAX_VALUE;
+    for (int j = 0; j < encoders.length; j++) {
+      for (int k = 0; k < 4; k++) {
+        if (vertices[inputLength][j][k] != null) {
+          Vector<ResultNode> edges = vertices[inputLength][j][k];
+          assert edges.size() == 1;
+          ResultNode result = edges.get(0);
+          if (result.getSize(true) < minimalSize) {
+            minimalSize = result.getSize(true);
+            minimalJ = j;
+            minimalK = k;
+          }
+        }
+      }
+    }
+    assert minimalJ != -1;
+    if (minimalJ >= 0) {
+      ResultNode result = vertices[inputLength][minimalJ][minimalK].get(0);
+      while (result.previous != null) {
+        result.previous.append(result);
+        result = result.previous;
+      }
+      return result;
+    } else {
+        throw new WriterException("Internal error: failed to encode");
+    }
+  }
+    
+  byte[] getBytesOfCharacter(int position,int charsetEncoderIndex) {
+    //TODO: Is there a more efficient way for a single character?
+    return stringToEncode.substring(position,position + 1).getBytes(encoders[charsetEncoderIndex].charset());
+ }
+
+  class ResultNode {
```

Comment:
private final?

Reply:
An instance of this class is returned as a result of calling MinimalEncoder.encode(). It has methods to retrieve the bits and the version that was chosen so that it can't be private as long as we call it from Encoder. 
I now made it final and made all members and method private but for these (they are called from Encoder):
void getBits(BitArray bits)
Version getVersion(ErrorCorrectionLevel ecLevel)
int getSize()
toString has to be public and it is called from the unit test.

Reply:
OK can it be static?

Reply:
Not currently because it accesses non static methods from the enclosing class which in turn access the input string and the encoders (the method ResultNode.getBits() for example needs to access the input).